### DISCUSSION TITLE
When are widgets dropped?

### DISCUSSION THREAD

**Discussion by AshfordN** [fltk-rs/fltk-rs#1002] [Q&A] [answered]
I know FLTK is a C++ library and so the memory underneath this rust binding is handled differently - according to C++ rules. However, according to Rust rules, objects are dropped when an object goes out of scope. So, in the following example, `label` should be dropped at the end of the `build_widget()` function. How then is the widget still valid, and when or how is the underlying memory freed? The `GroupExt::remove()` method says it removes a widget without deleting it, so if I were to make multiple calls to `build_widget()` followed by `win.remove_by_index(0)`, would this leak memory?

```rust
use fltk::{prelude::*, *};

fn build_widget() {
    let mut label = frame::Frame::default()
        .with_size(150, 50)
        .center_of_parent()
        .with_label("this is a test");
    label.set_frame(enums::FrameType::BorderBox);
}

fn main() {
    let root = app::App::default();

    let mut win = window::Window::default()
        .with_size(300, 300)
        .center_screen()
        .with_label("Sample");

    build_widget();

    win.end();
    win.show();

    root.run().unwrap();
}
```

**Comment by MoAlyousef** [comment, marked as answer]
Widgets are owned by their parents, so FLTK does its internal cleanup when a widget is no longer alive.
That means that there's no dropping on the Rust side. That's why you can spawn a window in function, and it won't be destroyed at the function's end of scope. Basically like most guis function.

It's also why this works:
```rust
let mut win = window::Window::default();
{
    let _label = frame::Frame::default();
    // The label isn't destroyed at the end of this scope, since it's now owned by the window. Also it's still accessible
}
win.end();
```
The label is still accessible thru the window's children, so the memory is technically not leaked since it's accessible.

Regarding remove(), it just removes the widget from it's current parent, it can still be added to another parent or the same parent as before. Of course you can create widgets outside of any GroupExt widget and purposfully leak them. 

To delete a widget, you can either call the clear method of its parent, or call the widget's delete function.

Technically even if you add a Drop impl which checks whether a widget has a parent or not, it would still interfere with the window's destruction.
